src/compiler/checker.ts

```diff
@@ -4420,12 +4420,19 @@ namespace ts {
             let container = getThisContainer(node, /*includeArrowFunctions*/ false);
             let parent = container && container.parent;
             if (parent && (isClassLike(parent) || parent.kind === SyntaxKind.InterfaceDeclaration)) {
-                if (!(container.flags & NodeFlags.Static)) {
+                if (!(container.flags & NodeFlags.Static) && !isConstructorParameter(node, container)) {
                     return getDeclaredTypeOfClassOrInterface(getSymbolOfNode(parent)).thisType;
                 }
             }
             error(node, Diagnostics.A_this_type_is_available_only_in_a_non_static_member_of_a_class_or_interface);
             return unknownType;
+
+            function isConstructorParameter(node: TypeNode, container: Node) {
+                if (container.kind === SyntaxKind.Constructor) {
+                    let ctor = (<ConstructorDeclaration>container);
+                    return !ctor.body.statements.some(st => st === node.parent);

```

Comment:
Also, the `some` method on `Array` is an ES5 thing. We restrict ourselves to ES3 so we can run anywhere. In this particular case I'd use the `forEach` helper function. But, as I said above, there's a better way to do the test.